core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/domain/solution/descriptor/SolutionDescriptor.java

```diff
@@ -927,6 +927,34 @@ private void visitAllEntities(Solution_ solution, Consumer<Object> visitor,
         }
     }
 
+    public void visitEntitiesByEntityClass(Solution_ solution, Class<?> entityClass, Consumer<Object> visitor,
+            Consumer<Collection<Object>> collectionVisitor) {
+        for (MemberAccessor entityMemberAccessor : entityMemberAccessorMap.values()) {
+            // TODO test this condition
+            if (entityClass.isAssignableFrom(entityMemberAccessor.getType())) {
+                Object entity = extractMemberObject(entityMemberAccessor, solution);
+                if (entity != null) {
+                    visitor.accept(entity);
+                }
+            }
+        }
+        for (MemberAccessor entityCollectionMemberAccessor : entityCollectionMemberAccessorMap.values()) {
+            // TODO simplify this (it's overly complex because it's intended for config validation)
+            Class<?> typeParameter = ConfigUtils.extractCollectionGenericTypeParameterLeniently(
+                    "FIXME", entityCollectionMemberAccessor.getDeclaringClass(),
+                    entityCollectionMemberAccessor.getType(),
+                    entityCollectionMemberAccessor.getGenericType(),
+                    null,
+                    entityCollectionMemberAccessor.getName());
+            // TODO test this condition
+            if (entityClass.isAssignableFrom(typeParameter)) {
+                Collection<Object> entityCollection = extractMemberCollectionOrArray(entityCollectionMemberAccessor, solution,
+                        false);
+                collectionVisitor.accept(entityCollection);
+            }
+        }
+    }
```

Comment:
The other part of the fix. This allows to only visit entities that are subclasses of the `entityClass` argument.